### PR TITLE
VersionId support in get,put object APIs

### PR DESCRIPTION
**Pull request by prakashsvmx**
VersionId support in get,put object APIs and respective response

### CODE REVIEW

**Review by ebozduman — CHANGES_REQUESTED** [review]

File: src/main/helpers.js [hunk 1]
```diff
@@ -362,4 +362,8 @@ export function extractMetadata(metaData) {
 export function getVersionId(headers={}){
   const versionIdValue = headers["x-amz-version-id"]
   return  versionIdValue || null
+}
+
+export function sanitizeETag(etag=''){
+  return etag.replace(/^"/g, '').replace(/"$/g, '').replace(/^&quot;/g, '').replace(/&quot;$/g, '').replace(/^&#34;/g, '').replace(/&#34;$/g, '')
```
[the remark this answers is not in the document]
Comment (prakashsvmx) [reply src/main/helpers.js:368]: Thank you @ebozduman  for the review . I will work on these.

**Comment by prakashsvmx** [comment]
@ebozduman  This helped me to figure out a minor bug in `fGetObject` with `versionId`  and i have fixed it and added the example as well. thank you. 
PTAL

File: src/test/functional/functional-tests.js [hunk 2]
```diff
@@ -1368,4 +1378,75 @@ describe('functional tests', function() {
...
+              versionId = res.versionId
+              done()
+            }else{
+              done(new Error('versionId not found in putObject response'))
+            }
+          })
+          .catch(done)
+      })
+
+      step(`getObject(bucketName, objectName, getOpts) :${versionedBucketName}, objectName:${versioned_100kbObjectName}`, done => {
```
Comment (ebozduman) [inline src/test/functional/functional-tests.js:1414]: I thought I gave this comment, but I cannot see it.

In all these `step` explanations, the standard is to have `_bucketName` before bucket name.

```suggestion
      step(`getObject(bucketName, objectName, getOpts)_bucketName :${versionedBucketName}, objectName:${versioned_100kbObjectName}`, done => {
```

There are several cases in your PR.

**Review by ebozduman — CHANGES_REQUESTED** [review]

File: src/test/functional/functional-tests.js (hunk 2 above)
Comment (prakashsvmx) [reply src/test/functional/functional-tests.js]: Thank you. I have applied the suggestion.

**Review by ebozduman — DISMISSED** [review]
LGTM

File: docs/API.md [hunk 3]
```diff
@@ -540,6 +606,20 @@ minioClient.fGetObject('mybucket', 'photo.jpg', '/tmp/photo.jpg', function(err)
   console.log('success')
 })
 ```
+
+__Example__
+To Stream a specific object version into a file.
+
+```js
+minioClient.fGetObject(bucketName, objNameValue, './download/MyImage.jpg}', {versionId:"03fd1247-90d9-4b71-a27e-209d484a234b"}, function(e) {
```
Comment (kanagarajkm) [inline docs/API.md:614]: `}` can be removed from `'./download/MyImage.jpg}'`

File: src/main/minio.js [hunk 4]
```diff
@@ -885,7 +908,7 @@ export class Client {
   // * `objectName` _string_: name of the object
   // * `filePath` _string_: file path of the file to be uploaded
   // * `metaData` _Javascript Object_: metaData assosciated with the object
-  // * `callback(err, etag)` _function_: non null `err` indicates error, `etag` _string_ is the etag of the object uploaded.
+  // * `callback(err, objInfo)` _function_: non null `err` indicates error, `objInfo` _object_ is the information about the object uploaded which containes versionId sting and etag string.
```
Comment (kanagarajkm) [inline src/main/minio.js:911]: `sting` -> `string`

Comment (kanagarajkm) [reply src/main/minio.js:911]: `containes` -> `contains`

Comment (kanagarajkm) [reply src/main/minio.js:911]: I feel this can be just `which contains versionId and etag`

File: src/test/functional/functional-tests.js [hunk 5]
```diff
@@ -35,6 +35,16 @@ try {
 } catch (err) {
   minio = require('minio')
 }
+let minIOHelpers
+try {
+  minIOHelpers = require('../../../dist/main/helpers')
```
Comment (kanagarajkm) [inline src/test/functional/functional-tests.js:40]: Is there a problem in using `../../main/helpers`

File: docs/API.md [hunk 6]
```diff
@@ -613,10 +702,19 @@ __Parameters__
...
+__Return Value__
+
+| Value  |  Type | Description  |
+|---|---|---|
+| `err` | _object_ | Error in case of any failures
+| `objInfo.etag`  | _string_  | `etag` of an object  |
+| `objInfo.versionId`  | _string_  | `versionId` of an object (optional)  |
+
 
 __Example__
```
Comment (kanagarajkm) [inline docs/API.md:720]: Example code snippet needs to be updated

Comment (prakashsvmx) [reply docs/API.md:723]: I have updated the snippet to reflect the same.

**Review by kanagarajkm — APPROVED** [review]

**Review by ebozduman — APPROVED** [review]
LGTM

**Review by kannappanr — APPROVED** [review]
LGTM

**Pull request merged**